### DISCUSSION TITLE
How to trigger and handle api-call related errors from server-action and toasts in one place in nextjs 13/14? Please share example or docs for reference

### DISCUSSION THREAD

**Discussion by manacy-keyvalue** [vercel/next.js#65681] [App Router]
I noticed the useFormState hook for this scenario but i cannot find an example where the error codes are mapped in a single place and toasts be shown in a common place as well.

I found a lot of related threads and react docs:
https://github.com/vercel/next.js/discussions/49426
https://pl.react.dev/reference/react-dom/hooks/useFormState

These don't have enough examples. And also couldn't find a way to reset this error once toast is closed without using a custom state and reset implementation in each component.

Please help! @timneutkens

**Comment by ghost** [comment]
# Approach 1 - Recreating a snackbar message via the key prop

Each time the form is submitted successfully, the snackbar is recreated via a `key` that is provided by `useActionState`. Internally, the snackbar can be dismissed using local state (via the `useState` hook). So the `key` provided by `useActionState` resets the internal dismissed state of the snackbar to `false` by recreating it. The downside of this approach is that if the key changes while the snackbar is still open, the exit transition of the old snackbar is not performed, but only the enter transition of the new snackbar. However, this approach most closely follows recommendations from the React docs regarding that [you might not need an effect](https://react.dev/learn/you-might-not-need-an-effect#resetting-all-state-when-a-prop-changes) and how to [reset state with a key](https://react.dev/reference/react/useState#resetting-state-with-a-key). Moreover, a local-state-only approach aligns nice with the dataflow and the component oriented architecture of the new paradigms since no external dependency to a Toaster or ContextProvider is present.

## Implementation:

Snackbar/Toast:
```jsx
export default function SuccessMessage({ messageId, message }: { messageId: number | string; message: string }) {
  // https://react.dev/learn/preserving-and-resetting-state#resetting-a-form-with-a-key
  // https://react.dev/learn/you-might-not-need-an-effect#resetting-all-state-when-a-prop-changes
  // https://react.dev/reference/react/useState#resetting-state-with-a-key
  // Reset the internal state and execute the enter transition/animation when a message is provided.
  return <Message key={messageId} message={message} />
}

function Message({ message }: { message: string }) {
  const [dismissed, setDismissed] = useState(false)

  const open = !dismissed && !!message

  const handleClose = (event: SyntheticEvent | Event, reason?: string) => {
    if (reason === 'clickaway') {
      return
    }
    setDismissed(true)
  }

  return (
    <Snackbar
      message={message}
      open={open}
      onClose={handleClose}
      autoHideDuration={4000}
      anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
    />
  )
}
``` 
Server Action:
```js
export async function updateAction(
  prevState: { messageId: number; message: string },
  formData: FormData
) {
  // do stuff
  return {
    messageId: prevState.messageId + 1,
    message: 'Successfully saved.'
  }
}
```

# Approach 2 - Show a toast via `useEffect`

A toast can be shown via `useEffect` by providing a return value (such as a timestamp) of `useActionState` in the dependency array of the effect. While this pattern is commonly used, it has a pitfall in strict mode (dev-only) when progressive enhancement leads to a hard page reload, and after such a page reload, the effect is being executed, showing a toast. When a component mounts, the effect runs twice in strict mode, leading to two toasts being shown for one form submission. This has to be considered in the implementation of the effect.

## Implementation
This is by far the best article I found so far concerning this implementation approach:
https://www.robinwieruch.de/next-forms/#toast-message-with-server-actions-in-next

# Approach 3 - Composing server actions with actions on the client

Alternatively, server actions can be composed with actions on the client: https://react.dev/reference/rsc/server-actions#composing-server-actions-with-actions

While this approach is perfectly legitimate according to the React docs, combined with `revalidatePath` and `redirect` from Next, there are still some subtle pitfalls.

When a server action that internally calls `revalidatePath` is awaited inside an action on the client, the state changes applied after the await statement are executed before the revalidated data is shown on the user interface.

When a server action that internally does a `redirect` is awaited inside an action on the client, `undefined` is returned to the client. However, the `redirect` function has a `never` return type. So, the code after the await statement on the client is still executed with the `undefined` return value that is not covered by Typescript, which might introduce bugs.

## Implementation
So, you can await the result of a server action inside an action on the client and then call a toast function. This client action is passed as a form's `action` prop. But be careful when using the `redirect` function in a Server Action.
